models/transactions.go

```diff
@@ -0,0 +1,137 @@
+package models
+
+import (
+	"database/sql"
+	"log"
+	"time"
+)
+
+type Transaction struct {
+	ID    string `json:"id"`
+	Lines []*TransactionLine
+}
+
+type TransactionLine struct {
+	AccountID string `json:"account"`
+	Delta     int    `json:"delta"`
+}
+
+func (t *Transaction) IsValid() bool {
+	if len(t.Lines) == 0 {
+		return false
+	}
+	sum := 0
+	for _, line := range t.Lines {
+		sum += line.Delta
+	}
+	if sum == 0 {
+		return true
+	}
+	return false
+}
+
+type TransactionDB struct {
+	DB *sql.DB
+}
+
+func (tdb *TransactionDB) IsExists(id string) bool {
+	var exists bool
+	err := tdb.DB.QueryRow("SELECT exists (SELECT * FROM transactions WHERE id=$1)", id).Scan(&exists)
+	if err != nil {
+		log.Println("Error executing transaction exists query:", err)
+	}
+	return exists
+}
+
+func (tdb *TransactionDB) DoTransaction(t *Transaction) bool {
+	// Start the transaction
+	var err error
+	txn, err := tdb.DB.Begin()
+	if err != nil {
+		log.Println("Error beginning transaction:", err)
+		return false
+	}
+
+	// Rollback transaction on any failures
+	defer func() {
+		if r := recover(); r != nil {
+			log.Fatal("Recovering while transaction", r)
+
+			log.Println("Rolling back the transaction:", t.ID)
+			err = txn.Rollback()
+			if err != nil {
+				log.Println("Error rolling back transaction:", err)
+			}
+		} else if err != nil {
+			log.Println("Rolling back the transaction:", t.ID)
+			err = txn.Rollback()
+			if err != nil {
+				log.Println("Error rolling back transaction:", err)
+			}
+		}
+	}()
+
+	// Accounts do not need to be predefined
+	// they are called into existence when they are first used.
+	accStmt, err := txn.Prepare(`INSERT INTO accounts (id) VALUES ($1)`)
+	if err != nil {
+		log.Println("Error preparing statement of accounts:", err)
+		return false
+	}
+	for _, line := range t.Lines {
+		var exists bool
+		err = txn.QueryRow("SELECT exists (SELECT * FROM accounts WHERE id=$1)", line.AccountID).Scan(&exists)
```

Comment:
Can we use Postgresql's Upsert instead of checking and inserting accounts entries?
https://www.postgresql.org/docs/9.5/static/sql-insert.html